teachopencadd/talktorials/T018_automated_cadd_pipeline/utils/helpers/dogsitescorer.py

```diff
@@ -0,0 +1,352 @@
+                file_extension = ".ccp4"
+
+            file_name = binding_site + file_extension
+            with open(output_path / file_name, "wb") as f:
+                f.write(response_file_content)
+    return
+
+
+def select_best_pocket(binding_site_df, selection_method, selection_criteria, ascending=False):
+    """
+    Select the best binding site from the table of all detected binding sites,
+    either by sorting the binding sites based on a set of properties in the table,
+    or by applying a function on the property values.
+
+    Parameters
+    ----------
+    binding_site_df : pandas.DataFrame
+        Binding site data retrieved from the DoGSiteScorer webserver.
+    selection_method : str
+        Selection method for selecting the best binding site.
+        Either 'sorting' or 'function'.
+    selection_criteria : str or list
+        If 'selection_method' is 'sorting':
+            List of one or more property names.
+        If 'selection_method' is 'function':
+            Any valid python syntax that generates a list-like object
+            with the same length as the number of detected binding sites.
+    ascending : bool
+        Optional; default: False.
+        If set to True, the binding site with the lowest value will be selected,
+        otherwise, the binding site with the highest value is selected.
+
+    Returns
+    -------
+    str
+        Name of the selected binding site.
+    """
+    df = binding_site_df
+
+    if selection_method == "sorting":
+        sorted_df = df.sort_values(by=selection_criteria, ascending=ascending)
+    elif selection_method == "function":
+        df["function_score"] = eval(selection_criteria)
+        sorted_df = df.sort_values(by="function_score", ascending=ascending)
+    else:
+        raise ValueError(f"Binding site selection method unknown: {selection_method}")
+
+    selected_pocket_name = sorted_df.iloc[0].name
+    return selected_pocket_name
+
+
+def calculate_pocket_coordinates_from_pocket_pdb_file(filepath):
+    """
+    Calculate the coordinates of a binding site using the binding site's PDB file
+    downloaded from DoGSiteScorer.
+
+    Parameters
+    ----------
+    filepath : str or pathlib.Path
+        Local filepath (including filename, without extension) of the binding site's PDB file.
+
+    Returns
+    -------
+    dict of list of int
+        Binding site coordinates in format:
+        `{'center': [x, y, z], 'size': [x, y, z]}`
+    """
+    with open(f"{filepath}.pdb") as f:
+        pdb_file_text_content = f.read()
+    pdb_file_df = pdb.load_pdb_file_as_dataframe(pdb_file_text_content)
+    pocket_coordinates_data = pdb_file_df["OTHERS"].loc[5, "entry"]
+    coordinates_data_as_list = pocket_coordinates_data.split(" ")
+    coordinates = []
+    for elem in coordinates_data_as_list:
+        try:
+            coordinates.append(float(elem))
+        except:
+            # FIXME specify exception
+            pass
```

Comment:
@Armin-Ariamajd, if `coordinates.append(float(elem))` throws an error, 
- what error are you expecting here and
- how do you want to handle it?

Reply:
It is supposed to throw errors;
the `coordinates_data_as_list` looks something like this:
```
['\tGeometric', 'pocket', 'center', 'at', '', '15.91', '', '32.33', '', '11.03', 'with', 'max', 'radius', '12.42']
```
And we only need the numbers. So we iterate over the list, and try to cast each element into a float, which will only work for numbers, whereas the strings raise a `ValueError` that are ignored by the `except: pass`. So we jump over the non-number elements of the list.

After the for-loop, `coordinates` looks like this:
```
[15.91, 32.33, 11.03, 12.42]
```
The first three numbers are the x-y-z coordinates of the center, and the fourth number is the radius. We assume a cubic pocket, and thus create the `pocket_coordinates` (which needs x-y-z coordinates of the center, plus length of the whole pocket in all three directions) from `coordinates` by:
```
pocket_coordinates = {
    "center": coordinates[:3],
    "size": [coordinates[-1] * 2 for dim in range(3)],
}
```
which will be the output of the function, and will look like this:
```
{'center': [15.91, 32.33, 11.03], 'size': [24.84, 24.84, 24.84]}
```